### PR TITLE
Use consistent lifetime and registration pattern for design-time services

### PR DESCRIPTION
**Pull request by AndriySvyryd**

### CODE REVIEW

File: src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -110,17 +80,21 @@ public static class DesignTimeServiceCollectionExtensions
...
-                .AddTransient(_ => context.GetService<IMigrationsAssembly>())
-                .AddTransient(_ => context.GetService<IMigrationsIdGenerator>())
-                .AddTransient(_ => context.GetService<IMigrationsModelDiffer>())
-                .AddTransient(_ => context.GetService<IMigrator>())
-                .AddTransient(_ => context.GetService<IRelationalTypeMappingSource>())
-                .AddTransient(_ => context.GetService<IDesignTimeModel>().Model)
-                .AddTransient(_ => context.GetService<IModelRuntimeInitializer>());
+        {
+            new EntityFrameworkRelationalServicesBuilder(services)
+                .TryAdd(context.GetService<IDatabaseProvider>())
```
Comment (bricelam) [inline src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs:85]: With `TryAdd`, do we need to invert the order the services are built? If a service is added by multiple components, the resolution order should be:

1. User design-time services
2. Extension design-time services
3. Provider design-time services
4. Base  design-time services

File: src/EFCore.SqlServer/Design/Internal/SqlServerDesignTimeServices.cs [hunk 2]
```diff
@@ -33,12 +26,15 @@ public class SqlServerDesignTimeServices : IDesignTimeServices
...
-                .AddSingleton<IAnnotationCodeGenerator, SqlServerAnnotationCodeGenerator>()
-                .AddSingleton<IDatabaseModelFactory, SqlServerDatabaseModelFactory>()
-                .AddSingleton<IModelValidator, SqlServerModelValidator>()
-                .AddSingleton<IProviderConfigurationCodeGenerator, SqlServerCodeGenerator>()
-                .AddSingleton<IRelationalTypeMappingSource, SqlServerTypeMappingSource>()
-                .AddSingleton<LoggingDefinitions, SqlServerLoggingDefinitions>();
+        {
+            Check.NotNull(serviceCollection, nameof(serviceCollection));
+
+            serviceCollection.AddEntityFrameworkSqlServer();
```
Comment (bricelam) [inline src/EFCore.SqlServer/Design/Internal/SqlServerDesignTimeServices.cs:32]: Does this change any scopes, do we need to add CreateScope calls somewhere now?

Comment (bricelam) [reply src/EFCore.SqlServer/Design/Internal/SqlServerDesignTimeServices.cs:32]: Also, do these conflict with the ones resolved from the DbContext services?

**Review by bricelam — COMMENTED** [review]
We should do a bit more testing to make sure this doesn't break existing apps, extensions, and providers.

File: src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs:85]: Assuming that User still uses the old registration pattern we would need to change `AddEntityFrameworkDesignTimeServices` to use the new pattern and then call it after `ConfigureProviderServices`.

I'll add tests for this.

File: src/EFCore.SqlServer/Design/Internal/SqlServerDesignTimeServices.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.SqlServer/Design/Internal/SqlServerDesignTimeServices.cs:32]: It does change lifetimes, many services were registered with the incorrect lifetime, that was the main driver behind this change.

It seems that there's an implicit scope now, but I could add explicit `CreateScope` calls for clarity.

DbContext services will be registered first, so this won't override them.

**Comment by AndriySvyryd** [comment]
This is ready for review

**Comment by ajcvickers** [comment]
@AndriySvyryd I'll take a look once @bricelam is happy.

**Review by bricelam — APPROVED** [review]

File: src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs (hunk 1 above)
Comment (bricelam) [reply src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs:88]: The tests look great. Thanks for adding them!

**Pull request merged**